File: commands.py
import re

class Command:
    def __init__(self, world, player):
        pass

class DebugCommand(Command):
    r = re.compile("^/debug (.+)$")

    def matches(self, s):
        m = self.r.match(s)
        if m:
            self.target = m[1]
            return True 
        else:
            return False

    def exec(self, player):
        count = 0
        try:
            self.target = int(self.target)
        except ValueError:
            pass
        for obj in player.world.items():
            if obj.name.lower() == self.target or obj.id == self.target:
                print("{}[id={}]: {}".format(obj.name, obj.id, obj))
                count += 1
        if count == 0:
            print(self.target + ": nessun oggetto con questo nome")

class QuitCommand(Command):
    def matches(self, s):
        if s.lower() in ["basta", "fine", "quit", "abbandona", "bye", "esci", "x"]:
            if len(s)>1:
                print("Basta che scrivi 'x' per uscire")
            return True
        return False
    
    def exec(self, player):
        print("Alla prossima!")

class LookCommand(Command):
    r = re.compile("^(guarda|osserva|ispeziona) (.+)$")

    def matches(self, s):
        self.target = None
        if s in ["guarda", "guarda in giro", "guardati in giro", 
            "guardati intorno", "guardati aggiro", "osserva", "ispeziona"]:
            return True
        m = self.r.match(s)
        if m:
            target = m[2]
            self.target = target
            return True
        return False
    
    def exec(self, player):
        if self.target is None:
            player.world.look_at(player, player.container)
        else:
            for obj in player.container.items() + player.items():
                if obj.name.lower() == self.target:
                    player.world.look_at(player, obj)
                    return
            print("Non vedo nessun", self.target)

class GoCommand(Command):
    r = re.compile(r'^(vai a|vai verso|vai) (.*)$')

    def direction(self, s):
        d = {
            'n': ['nord'],
            's': ['sud'],
            'e': ['est'],
            'w': ['o','ovest'],
            'u': ['a','alto',"l'alto",'su','sù'],
            'd': ['b','basso',"il basso",'giu','giù'],
        }
        for key, vals in d.items():
            if s==key or s in vals:
                return key
        return None

    def matches(self, s):
        self.dir = self.direction(s)
        if self.dir is not None:
            return True
        m = self.r.match(s)
        if m:
            self.dir = self.direction(m[2])
            if self.dir is not None:
                print("Basta che dici:",m[2])
                if len(m[2]) > 1:
                    print("...oppure anche solo:", self.dir)
            else:
                print(m[2] + " non è una direzione valida. Prova: n,nord,s,sud,e,est,w,ovest,u,alto,d,basso")
            return True
        return False

    def exec(self, player):
        if self.dir is None:
            return
        if not player.container.passages:
            print("non c'è nessun passaggio")
            return
        try:
            room_id = player.container.passages[self.dir]
            player['container_id'] = room_id
        except KeyError:
            print("impossibile muoversi in questa direzione")
            return

class TakeCommand(Command):
    r = re.compile(r'^(prendi|raccogli|afferra) (.+)$')

    def matches(self, s):
        m = self.r.match(s)
        if m:
            self.target = m[2]
            return True
        return False
    
    def exec(self, player):
        for obj in player.container.visible_items() + player.visible_items():
            if obj == player:
                continue
            if obj.name.lower() == self.target:
                if not obj.be("collectable"):
                    print("non puoi sollevare", obj.name)
                elif not obj.is_reachable_by(player):
                    print("non puoi raggiungere", obj.name)
                else:
                    obj["container_id"] = player.id
                    print("Ho preso", obj.name)
                return
        print("Non vedo nessun", self.target)

class DropCommand(Command):
    r = re.compile(r'^(lascia|molla) (.+)$')
    r2 = re.compile(r'^(metti|lascia|inserisci) (.+) (dentro|in|nel|nello|nella|nell\'|sull\'|sul|nella|sullo|sotto|sopra) (.+)$')

    def matches(self, s):
        m = self.r2.match(s)
        if m:
            self.target = m[2]
            self.box = m[4]
            return True
        m = self.r.match(s)
        if m:
            self.target = m[2]
            self.box = None
            return True
        return False
    
    def exec(self, player):
        for obj in player.items():
            if obj.name.lower() == self.target:
                assert obj.be("collectable")
                break
        else:
            print("Non hai", self.target)
            return
        if self.box:
            for box in player.visible_items() + player.container.visible_items():
                if box.name.lower() == self.box:
                    if not box.is_reachable_by(player):
                        print(box.name, "irraggiungibile")
                        return
                    if not box.be("container"):
                        print ("non puoi mettere", obj.name, "dentro", box.name)
                        return
                    if box.closed:
                        print(box.name, "è chiuso")
                        return    
                    if box == obj:
                        print("stai provando a rompere lo spaziotempo,non va bene,sciocco hobbit")
                        return
                    break
            else:
                print("non vedo", self.box)
                return
            obj["container_id"] = box.id
            print("hai messo", obj.name, "dentro", box.name)     
        else:
            obj["container_id"] = player.container.id
            print("hai lasciato", obj.name)                     

class SayCommand(Command):
    r = re.compile(r'^(di|dì|pronuncia|urla|sussurra|bisbiglia)(:| ) *["\']?(.+?)["\']?$')

    def matches(self, s):
        m=self.r.match(s)
        if m:
            self.phrase=m[3]
            return True
        return False

    def exec(self, player):
        print(player.name+ ":", self.phrase)

class OpenCommand(Command):
    r = re.compile(r'^(apri|sblocca) (.+)$')

    def matches(self, s):
        m = self.r.match(s)
        if m:
            self.target = m[2]
            return True
        return False
    
    def exec(self, player):
        for obj in player.container.visible_items() + player.visible_items():
            if obj.name.lower() == self.target:
                if not obj.is_reachable_by(player):
                    print("non riesco a raggiungere", obj.name)
                    return
                if not obj.be("openable"):
                    print("non puoi aprire", obj.name)
                    return
                if not obj.closed:
                    print("sciocco hobbit", obj.name, "è già aperto")
                    return
                if obj.locked:
                    for o in player.items():
                        if obj == o.opens :
                            obj["locked"] = False
                            obj["closed"] = False
                            print("hai aperto", obj.name, "con", o.name)
                            break
                    else:
                        print("non hai la chiave per aprire", obj.name)
                        return
                else:
                    obj["closed"] = False
                    print("hai aperto", obj.name)
                lst = obj.items()
                if lst:
                    print("contiene:")
                    for x in lst:
                        print("-", x.name)
                else:
                    print("è vuoto")
                return
        print("Non vedo nessun", self.target)

class EnterCommand(Command):
    r = re.compile(r'^(entra) (in|nel|nell\'|nello|nella|dentro) (.+)$')

    def matches(self, s):
        m = self.r.match(s)
        if m:
            self.target = m[3]
            return True
        return False
    
    def exec(self, player):
        for obj in player.container.visible_items() + player.visible_items():
            if obj.name.lower() == self.target:
                if not obj.is_reachable_by(player):
                    print("non arrivi a", obj.name)
                    return
                if obj.container == player:
                    print("per problemi di manutenzione non puoi entrare in un'oggetto che hai in mano, perfavore lasciare l'oggetto in cui si desidera entrare")
                    return
                if not obj.be("container"):
                    print("non puoi entrare dentro", obj.name)
                    return
                if obj.closed:
                    print(obj.name, "è chiuso")
                    return
                else:
                    player["container_id"] = obj.id
                    print("sei entrato dentro", obj.name)
                    return
        else:
            print("non vedo", self.target)

class ExitCommand(Command):
    r = re.compile(r'^(esci|emergi) (da|dal|dalla|dallo) (.+)$')

    def matches(self, s):
        m = self.r.match(s)
        if m:
            self.target = m[3]
            return True
        return False
    
    def exec(self, player):
        if self.target != player.container.name.lower():
            print("non sei dentro", self.target)
            return
        if player.container.container:
            if player.container.closed:
                print(player.container.name, "è chiuso")
                return
            print("sei uscito da", player.container.name)
            player["container_id"] = player.container["container_id"]
        else:
            if player.container.passages:
                print("devi dire una direzione per uscire")
            else:
                print("non puoi uscire :( !")     

        
Commands = [
    DebugCommand,
    QuitCommand,
    LookCommand,
    GoCommand,
    TakeCommand,
    DropCommand,
    SayCommand,
    OpenCommand,
    EnterCommand,
    ExitCommand,
]

File: commands.js

function quit(input) {
    input.toLowerCase();
    input.trim();
    if (["basta", "fine", "quit", "abbandona", "bye", "esci", "x"].includes(input)) {
        return (player, world) => {
            if (input.length > 1) world.print("Basta che scrivi 'x' per uscire");
            world.print("Alla prossima!");
            return true;
        };
    }
    return null;
}

function debug(input) {
    const re = /^debug (.+)$/;
    let m = re.exec(input);
    if (m) {
        let target_name = m[1];
        return (player, world) => {
            let target = null;
            let target_id = parseInt(target_name);
            if (target_id >= 0) {
                target = world.item(target_id);            
            } else {
                target = world.item_by_name(target_name);
            }
            if (target) {
                world.print(`${target.name}=${JSON.stringify(target)}`);
            } else {
                world.print(`${target_name}: nessun oggetto con questo nome`);
            }
        }
    }
    return null;
}

function look(input) {
    if (["guarda", "guarda in giro", "guardati in giro",
        "guardati intorno", "guardati aggiro", "osserva", "ispeziona"].includes(input)) {
        return (player, world) => {
            world.look_at(player, world.item(player.container_id));
        };
    }
    const re = /^(guarda|osserva|ispeziona) (.+)$/;
    let m = re.exec(input);
    if (m) {
        let target_name = m[2];
        return (player, world) => {
            let items = world.items_in(player.container_id).concat(world.items_in(player));
            const target = world.item_by_name(target_name, items);
            if (target) world.look_at(player, target);
            else world.print(`Non vedo nessun ${target_name}`);
        }
    }
}

function go(input) {

    function direction(s) {
        const match = Object.entries({
            'n': ['nord'],
            's': ['sud'],
            'e': ['est'],
            'w': ['o','ovest'],
            'u': ['a','alto',"l'alto",'su','sù'],
            'd': ['b','basso',"il basso",'giu','giù'],
        }).filter(pair => (pair[0] == s || pair[1].includes(s)));
        if (match.length > 0) return match[0][0];
        return null;
    }

    function move(player, world, direction) {
        const passages = world.item(player.container_id).passages;
        if (passages) {
            const room_id = passages[direction];
            if (room_id) {
                player.container_id = room_id;
            } else {
                world.print(`impossibile muoversi in questa direzione`);
            }
        } else {
            world.print(`non c'è nessun passaggio`);
        }
    }

    let dir = direction(input);
    if (dir === null) {
        const r = /^(vai a|vai verso|vai) (.*)$/;
        const m = r.exec(input);
        if (m) {
            dir = direction(m[2]);
            if (dir) {
                return (player, world) => {
                    world.print(`Basta che dici: ${m[2]}`);
                    if (m[2].length > 1) world.print(`...o anche solo: ${dir}`);
                    move(player, world, dir);
                }
            } else {
                return (player, world) => {
                    world.print(`${m[2]} non è una direzione valida. Prova: n,nord,s,sud,e,est,w,ovest,u,alto,d,basso`);
                }
            }
        }
    } else {
        return (player, world) => {move(player, world, dir);}
    }
}

function take(input) {
    const r = /^(prendi|raccogli|afferra) (.+)$/;
    const m = r.exec(input);
    if (m) {
        const target_name = m[2];
        return (player, world) => {
            const item = world.item_by_name(target_name,
                world.visible_items_in(player.container_id).concat(
                world.visible_items_in(player)));
            if (!item) {
                world.print(`non vedo nessun ${target_name}`);
                return;
            }
            if (!(item.attributes && item.attributes.includes("collectable"))) {
                world.print(`non puoi sollevare ${item.name}`);
            } else if (!world.is_reachable_by(item, player)) {
                world.print(`non puoi raggiungere ${item.name}`);
            } else {
                item.container_id = player.id;
                world.print(`ho preso ${item.name}`);
            }
        }
    }
}

function put(input) {
    function check_have(target_name, player, world) {
        const item = world.item_by_name(target_name, world.items_in(player));
        if (!item) {
            world.print(`non hai ${target_name}`);
            return null;
        } else if (item.attributes.includes("collectable")) {
            return item;
        } else {
            world.print(`non puoi sollevare ${target_name}`);
            return null;
        }
    }

    const r2 = /^(metti|lascia|inserisci) (.+) (dentro|in|nel|nello|nella|nell\'|sull\'|sul|nella|sullo|sotto|sopra) (.+)$/;
    const m2 = r2.exec(input);

    if (m2) {
        const target_name = m2[2];
        const box_name = m2[4];
        return (player, world) => {
            const item = check_have(target_name, player, world);
            const box = world.item_by_name(box_name, 
                world.visible_items_in(player).concat(
                world.visible_items_in(player.container_id)));
            if (item) {
                if (!box) {
                    world.print(`non vedo ${box_name}`);
                } else if (!world.is_reachable_by(box, player)) {
                    world.print(`${box.name} irraggiungibile`);
                } else if (!box.attributes.includes("container")) {
                    world.print(`non puoi mettere ${item.name} dentro ${box.name}`);
                } else if (box.closed) {
                    world.print(`${box.name} è chiuso`);
                } else if (box.id === item.id) {
                    world.print(`Stai provando a rompere lo spaziotempo. Sciocco hobbit!`);
                } else {
                    item.container_id = player.container_id;
                    world.print(`hai messo ${item.name} dentro ${box.name}`);
                }
            }
        }
    }
    const r = /^(lascia|molla) (.+)$/;
    const m = r.exec(input);
    if (m) {
        const item_name = m[2];
        return (player, world) => {
            const item = world.item_by_name(item_name, world.items_in(player));
            if (item) {
                item.container_id = player.container_id;
                world.print(`hai lasciato ${item.name}`);
            } else {
                world.print(`non hai ${item_name}`);
            }
        }
    }
}

function say(input) {
    const r = /^(di|dì|pronuncia|urla|sussurra|bisbiglia)(:| ) *["\']?(.+?)["\']?$/;
    const m = r.exec(input);
    if (m) {
        const phrase = m[3];
        return (player, world) => {
            world.print(`${player.name}: ${phrase}`);
        }
    }
}

function open(input) {
    const r = /^(apri|sblocca) (.+)$/;
    const m = r.exec(input);
    if (m) {
        const item_name = m[2];
        return (player, world) => {
            const item = world.item_by_name(item_name,
                world.visible_items_in(player.container_id).concat(
                    world.visible_items_in(player)));
            function done(item) {
                const items = world.items_in(item);
                if (items.length>0) {
                    world.print(`contiene:`);
                    items.forEach(obj => {
                        world.print(`- ${obj.name}`);
                    });
                } else {
                    world.print(`è vuoto!`);
                }
            }
            if (!item) {
                world.print(`non vedo nessun ${item_name}`);
            } else if (!world.is_reachable_by(item, player)) {
                world.print(`non riesco a raggiungere ${item.name}`);
            } else if (!item.attributes.includes("openable")) {
                world.print(`non puoi aprire ${item.name}`);
            } else if (!item.closed) {
                world.print(`sciocco hobbit, ${item.name} è già aperto`);
            } else if (item.locked) {
                const keys = world.items_in(player).filter(obj => (obj.opens_id === item.id));
                if (keys.length>0) {
                    item.locked = false;
                    item.closed = false;
                    world.print(`hai aperto ${item.name} con ${keys[0].name}`);
                    done(item);
                } else {
                    world.print(`non hai la chiave per aprire ${item.name}`);
                }
            } else {
                item.closed = false;
                world.print(`hai aperto ${item.name}`);
                done(item);
            }
        }
    }
}

function enter(input) {
    const r = /^(entra) (in|nel|nell\'|nello|nella|dentro) (.+)/;
    const m = r.exec(input);
    if (m) {
        const obj_name = m[3];
        return (player, world) => {
            const obj = world.item_by_name(obj_name, 
                world.visible_items_in(player.container_id).concat(
                    world.visible_items_in(player)));
            if (!obj) {
                world.print(`non vedo ${obj_name}`);
            } else if (!world.is_reachable_by(obj, player)) {
                world.print(`non arrivi a ${obj.name}`);
            } else if (obj.container_id === player.id) {
                world.print(`per problemi di manutenzione non puoi entrare in un'oggetto che hai in mano, perfavore lasciare l'oggetto in cui si desidera entrare`);
            } else if (!obj.attributes.includes("container")) {
                world.print(`non puoi entrare dentro ${obj.name}`);
            } else if (obj.closed) {
                world.print(`${obj.name} è chiuso`);
            } else {
                player.container_id = obj.id;
                world.print(`sei entrato dentro ${obj.name}`);
            }
        }
    }
}

function exit(input) {
    const r = /^(esci|emergi) (da|dal|dalla|dallo) (.+)$/;
    const m = r.exec(input);
    if (m) {
        const container_name = m[3];
        return (player, world) => {
            const container = world.item(player.container_id);
            if (!world.has_name(container, container_name)) {
                world.print(`non sei dentro ${container_name}`);
            } else if (container.container_id) {
                if (container.closed) {
                    world.print(`${container.name} è chiuso`);
                } else {
                    player.container_id = container.container_id;
                    world.print(`sei uscito da ${container.name}`);
                }
            } else if (container.passages) {
                world.print(`devi dire una direzione per uscire`);
            } else {
                world.print(`apparentemente sei in un luogo senza uscite!`);
            }
        }
    }
}

exports.commands = [quit, debug, look, go, take, put, say, open, enter, exit];